### PR TITLE
Fungibles adapter and MultiLocation::X8

### PR DESCRIPTION
**Pull request by gavofyork**
Basic implementation allows the Assets pallet to be used in XCM. Some additional adapters/converters should allow more non-native assets to be sent & received.

Requires paritytech/substrate#8425.

This also includes some minor alterations to CI mainly for the renaming of a few of the labels.

### CODE REVIEW

File: xcm/src/v0/multi_location.rs [hunk 1]
```diff
@@ -56,6 +56,14 @@ pub enum MultiLocation {
 	X3(Junction, Junction, Junction),
 	/// A relative path comprising four junctions.
 	X4(Junction, Junction, Junction, Junction),
+	/// A relative path comprising four junctions.
```
Comment (xlc) [inline xcm/src/v0/multi_location.rs:59]: looks like we need some macro to generate those?

File: xcm/xcm-builder/src/fungibles_adapter.rs [hunk 2]
```diff
@@ -0,0 +1,257 @@
+// Copyright 2020 Parity Technologies (UK) Ltd.
```
Comment (shaunxw) [inline xcm/xcm-builder/src/fungibles_adapter.rs:1]:
```suggestion
// Copyright 2021 Parity Technologies (UK) Ltd.
```

File: xcm/xcm-builder/src/fungibles_adapter.rs [hunk 3]
```diff
@@ -0,0 +1,257 @@
...
+		Ok((what, amount))
+	}
+}
+
+pub struct FungiblesAdapter<Assets, Matcher, AccountIdConverter, AccountId>(
+	PhantomData<(Assets, Matcher, AccountIdConverter, AccountId)>
+);
+
+impl<
+	Assets: Fungibles<AccountId>,
```
Comment (shaunxw) [inline xcm/xcm-builder/src/fungibles_adapter.rs:230]: Only assets `deposit` and `withdraw` is used in in this impl, I think we could use a minimum trait bound here instead including only these two fn. In this way `FungiblesAdapter` would be easier to be reused with other assets impl like `orml-tokens`, without implementing all methods in `Fungibles`.

File: xcm/src/v0/multi_location.rs (hunk 1 above)
Comment (gavofyork) [reply xcm/src/v0/multi_location.rs:59]: yeah, that'd be nice

File: xcm/xcm-builder/src/fungibles_adapter.rs (hunk 3 above)
Comment (gavofyork) [reply xcm/xcm-builder/src/fungibles_adapter.rs:230]: with the possible exception of `transfer`, all will be needed for XCM eventually.

**Comment by gavofyork** [comment]
cc @s3krit

**Review by s3krit — APPROVED** [review]
CI/label changes look good

File: xcm/xcm-builder/src/fungibles_adapter.rs [hunk 4]
```diff
@@ -0,0 +1,267 @@
...
+	}
+}
+
+/// Generic third-party conversion trait. Use this when you don't want to force the user to use default
+/// impls of `From` and `Into` for the types you wish to convert between.
+///
+/// One of `convert`/`convert_ref` and `reverse`/`reverse_ref` MUST be implemented. If possible, implement
+/// `convert_ref`, since this will never result in a clone. Use `convert` when you definitely need to consume
+/// the source value.
+pub trait Convert<A: Clone, B: Clone> {
```
Comment (apopiak) [inline xcm/xcm-builder/src/fungibles_adapter.rs:54]: DQ: What's the difference/motivation relative to `TryFrom` (see docs [here](https://doc.rust-lang.org/std/convert/trait.TryFrom.html))?

Comment (gavofyork) [reply xcm/xcm-builder/src/fungibles_adapter.rs:54]: `TryFrom` is intrinsic to the (source) type in question and as such it is intended for the situation where there is and could only be one sensible way of converting from one to the other type.

In cases where there may be several reasonable ways of converting between two types, then a third-party "converter" type is needed which is able to select/define the means of conversion.

File: xcm/src/v0/multi_location.rs (hunk 1 above)
Comment (shawntabrizi) [reply xcm/src/v0/multi_location.rs]: https://github.com/paritytech/polkadot/issues/2750

File: xcm/xcm-builder/src/fungibles_adapter.rs [hunk 5]
```diff
@@ -0,0 +1,267 @@
...
+/// Implementation of `Convert<Vec<u8>, _>` using the parity scale codec.
+pub struct Decoded;
+impl<T: Clone + Encode + Decode> Convert<Vec<u8>, T> for Decoded {
+	fn convert_ref(bytes: impl Borrow<Vec<u8>>) -> result::Result<T, ()> {
+		T::decode(&mut &bytes.borrow()[..]).map_err(|_| ())
+	}
+	fn reverse_ref(value: impl Borrow<T>) -> result::Result<Vec<u8>, ()> { Ok(value.borrow().encode()) }
+}
+
+/// Converter struct implementing `AssetIdConversion` converting a numeric asset ID (must by TryFrom/TryInto<u128>)
```
Comment (shawntabrizi) [inline xcm/xcm-builder/src/fungibles_adapter.rs:146]:
```suggestion
/// Converter struct implementing `AssetIdConversion` converting a numeric asset ID (must be TryFrom/TryInto<u128>)
```

**Review by shawntabrizi — APPROVED** [review]

**Pull request merged**